build/build_utils.py

```diff
@@ -0,0 +1,306 @@
+        os.chmod(path, stat.S_IWUSR)
+        func(path)
+        print_debug('rmtree:read-only file/path (%s)' % (path, ))
+    else:
+        raise
+
+class ordered_set(set):
+    def __init__(self):
+        set.__init__(self)
+        self.__list = list()
+
+    def add(self, o):
+        if not o in self:
+            set.add(self, o)
+            self.__list.append(o)
+
+    def __iter__(self):
+        return self.__list.__iter__()
+
+class Tarball(object):
+    def unpack(self):
+        print_log('Extracting %s to %s' % (self.archive_file, self.builder.working_dir))
+
+        os.makedirs(self.build_dir)
+        self.builder.exec_msys([self.builder.tar, 'ixf', convert_to_msys(self.archive_file), '-C', convert_to_msys(self.build_dir), '' if self.tarbomb else '--strip-components=1'])
+
+        print_log('Extracted %s' % (self.archive_file,))
+
+class MercurialRepo(object):
+    def unpack(self):
+        print_log('Cloning %s to %s' % (self.repo_url, self.build_dir))
+        self.exec_cmd('hg clone %s %s-tmp' % (self.repo_url, self.build_dir))
+        shutil.move(self.build_dir + '-tmp', self.build_dir)
+        print_log('Cloned %s to %s' % (self.repo_url, self.build_dir))
+
+    def update_build_dir(self):
+        print_log('Updating directory %s' % (self.build_dir,))
+        self.exec_cmd('hg pull -u', working_dir=self.build_dir)
+
+class GitRepo(object):
+    def unpack(self):
+        print_log('Cloning %s to %s' % (self.repo_url, self.build_dir))
+
+        self.builder.exec_msys('git clone %s %s-tmp' % (self.repo_url, self.build_dir))
+        shutil.move(self.build_dir + '-tmp', self.build_dir)
+
+        if self.fetch_submodules:
+            self.builder.exec_msys('git submodule update --init',  working_dir=self.build_dir)
+
+        if self.tag:
+            self.builder.exec_msys('git checkout -f %s' % self.tag, working_dir=self.build_dir)
+
+        print_log('Cloned %s to %s' % (self.repo_url, self.build_dir))
+
+    def update_build_dir(self):
+        print_log('Updating directory %s' % (self.build_dir,))
+
+        # I don't like too much this, but at least we ensured it is properly cleaned up
+        self.builder.exec_msys('git clean -xdf', working_dir=self.build_dir)
+
+        if self.tag:
+            self.builder.exec_msys('git fetch origin', working_dir=self.build_dir)
+            self.builder.exec_msys('git checkout -f %s' % self.tag, working_dir=self.build_dir)
+        else:
+            self.builder.exec_msys('git checkout -f', working_dir=self.build_dir)
+            self.builder.exec_msys('git pull --rebase', working_dir=self.build_dir)
+
+        if self.fetch_submodules:
+            self.builder.exec_msys('git submodule update --init', working_dir=self.build_dir)
+
+        if os.path.exists(self.patch_dir):
+            print_log("Copying files from %s to %s" % (self.patch_dir, self.build_dir))
+            self.builder.copy_all(self.patch_dir, self.build_dir)
+
+class Project(object):

```

Comment:
Single file with the Project class, right (and also the Tool one, these are the base building blocks ...) ? 

The Meson and CmakeProject where should go, in a base_builders.py module ?

Thanks